internal/controller/apiextensions/composite/composed.go

```diff
@@ -39,16 +39,42 @@ import (
 
 // Error strings
 const (
-	errApply       = "cannot apply composed resource"
-	errFetchSecret = "cannot fetch connection secret"
-	errReadiness   = "cannot check whether composed resource is ready"
-	errUnmarshal   = "cannot unmarshal base template"
-	errFmtPatch    = "cannot apply the patch at index %d"
-	errGetSecret   = "cannot get connection secret of composed resource"
-	errNamePrefix  = "name prefix is not found in labels"
-	errName        = "cannot use dry-run create to name composed resource"
+	errGetComposed   = "cannot get composed resource"
+	errGCComposed    = "cannot garbage collect composed resource"
+	errApplyComposed = "cannot apply composed resource"
+	errFetchSecret   = "cannot fetch connection secret"
+	errReadiness     = "cannot check whether composed resource is ready"
+	errUnmarshal     = "cannot unmarshal base template"
+	errGetSecret     = "cannot get connection secret of composed resource"
+	errKindChanged   = "cannot change the kind of an existing composed resource"
+	errNamePrefix    = "name prefix is not found in labels"
+	errName          = "cannot use dry-run create to name composed resource"
+	errMixed         = "cannot mix named and anonymous resource templates"
+	errDuplicate     = "resource template names must be unique within their Composition"
+
+	errFmtPatch      = "cannot apply the patch at index %d"
+	errFmtRenderIdx  = "cannot render composed resource at index %d"
+	errFmtRenderName = "cannot render composed resource template %q"
 )
 
+// Annotation keys.
+const (
+	// TODO(negz): This doesn't belong here.
+	AnnotationKeyCompositionTemplateName = "crossplane.io/composition-template-name"
+)
+
+// SetCompositionTemplateName sets the name of the composition template used to
+// reconcile a composed resource as an annotation.
+func SetCompositionTemplateName(o metav1.Object, name string) {
+	meta.AddAnnotations(o, map[string]string{AnnotationKeyCompositionTemplateName: name})
+}
+
+// GetCompositionTemplateName gets the name of the composition template used to
+// reconcile a composed resource from its annotations.
+func GetCompositionTemplateName(o metav1.Object) string {
+	return o.GetAnnotations()[AnnotationKeyCompositionTemplateName]
+}
```

Comment:
I considered persisting the template name in the XR's resourceRefs array, rather than as an annotation on the composed resource. Technically doing so would be a violation of API conventions, which require that references be "a strict subset" of the ObjectReference type.

I feel like there was another reason to persist the name to the actual resource rather than the reference, but it's not immediately coming to mind.